codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.github.luben.zstd.Zstd;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ChannelPromiseNotifier;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.handler.codec.EncoderException;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.ObjectUtil;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeUnit;
+import java.util.zip.Adler32;
+
+
+import static io.netty.handler.codec.compression.ZstdConstants.DEFAULT_BLOCK_SIZE;
+import static io.netty.handler.codec.compression.ZstdConstants.MAX_BLOCK_SIZE;
+import static io.netty.handler.codec.compression.ZstdConstants.MIN_BLOCK_SIZE;
+import static io.netty.handler.codec.compression.ZstdConstants.COMPRESSION_LEVEL_BASE;
+import static io.netty.handler.codec.compression.ZstdConstants.HEADER_LENGTH;
+import static io.netty.handler.codec.compression.ZstdConstants.DEFAULT_COMPRESS_LEVEL;
+import static io.netty.handler.codec.compression.ZstdConstants.BLOCK_TYPE_NON_COMPRESSED;
+import static io.netty.handler.codec.compression.ZstdConstants.BLOCK_TYPE_COMPRESSED;
+import static io.netty.handler.codec.compression.ZstdConstants.ZSTD_NUMBER;
+import static io.netty.handler.codec.compression.ZstdConstants.TOKEN_OFFSET;
+import static io.netty.handler.codec.compression.ZstdConstants.COMPRESSED_LENGTH_OFFSET;
+import static io.netty.handler.codec.compression.ZstdConstants.DECOMPRESSED_LENGTH_OFFSET;
+import static io.netty.handler.codec.compression.ZstdConstants.DEFAULT_CHECKSUM_OFFSET;
+import static io.netty.util.internal.ThrowableUtil.unknownStackTrace;
+
+/**
+ *  Compresses a {@link ByteBuf} using the Zstandard algorithm.
+ *  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *     * * * * * * * * * * * * *
+ *  * ZSTD number * CompressType *  Compressed *  Decompressed *  Checksum *  +  *  ZSTD compressed *
+ *  *       *              *    length   *     length    *           *     *      block      *
+ *  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *     * * * * * * * * * * * * *
+ */
```

Comment:
This javadoc comment renders quite poorly. Try reformatting it using the javadoc HTML support, or something.
Also, please add some text to accompany the diagram, to make it clear to the reader what they are looking at.

Reply:
I've done it